test_report/src/report.rs

```diff
@@ -0,0 +1,268 @@
+            version: std::env!("CARGO_PKG_VERSION").to_string(),
+            upload_time: None,
+        });
+        Self(RefCell::new(TestReport {
+            test_result: TestResult::default(),
+        }))
+    }
+
+    fn serialize_test_result(&self) -> Vec<u8> {
+        let test_result = self.0.borrow().test_result.clone();
+        let buf: Vec<u8> = prost::Message::encode_to_vec(&test_result);
+        buf
+    }
+
+    // sends out to the trunk api
+    pub fn publish(&self) -> bool {
+        let path = self.save();
+        if path.is_err() {
+            return false;
+        }
+        let resolved_path = path.unwrap_or_default();
+        let token = std::env::var("TRUNK_API_TOKEN").unwrap_or_default();
+        let org_url_slug = std::env::var("TRUNK_ORG_URL_SLUG").unwrap_or_default();
+        if token.is_empty() || org_url_slug.is_empty() {
+            println!("Token or org url slug not set");
+            return false;
+        }
+        if let Some(uploader_metadata) = &mut self.0.borrow_mut().test_result.uploader_metadata {
+            uploader_metadata.upload_time = Some(Timestamp {
+                seconds: Utc::now().timestamp(),
+                nanos: Utc::now().timestamp_subsec_nanos() as i32,
+            });
+        }
+        let upload_args = trunk_analytics_cli::upload::UploadArgs::new(
+            token,
+            org_url_slug,
+            vec![resolved_path],
+            // TODO - this isn't right, find the repo root
+            "..".to_string(),
+        );
+        match tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .unwrap()
+            .block_on(trunk_analytics_cli::upload::run_upload(
+                upload_args,
+                None,
+                None,
+                None,
+                None,
+            )) {
+            Ok(_) => return true,
+            Err(e) => {
+                println!("Error uploading: {:?}", e);
+                return false;
+            }
+        }
+    }
+
+    // saves to local fs and returns the path
+    pub fn save(&self) -> Result<String, anyhow::Error> {
+        let buf = self.serialize_test_result();
+        if let Ok(named_temp_file) = tempfile::Builder::new().suffix(".bin").tempfile() {
+            std::fs::write(&named_temp_file, buf).unwrap_or_default();
+            let (_, path) = named_temp_file.keep().unwrap();
+            let path_str = path.to_str().unwrap_or_default();
+            return Ok(path_str.to_string());
+        }
+        Err(anyhow::anyhow!("Error saving test report"))
+    }
+
+    // adds a test to the test report
+    pub fn add_test(
+        &self,
+        id: Option<String>,
+        name: String,
+        classname: String,
+        file: String,
+        parent_name: String,
+        line: Option<i32>,
+        status: Status,
+        attempt_number: i32,
+        started_at: i64,
+        finished_at: i64,
+        output: String,
```

Comment:
Normally I'd prefer to use an object here, but `Magnus` was making this difficult with the optional fields. Using the values directly reduces boilerplate needed and makes it easier to use in ruby.